src/lib/mdns/minimal/core/QName.cpp

```diff
@@ -134,5 +134,22 @@ const uint8_t * SerializedQNameIterator::FindDataEnd()
     return nullptr;
 }
 
+bool SerializedQNameIterator::operator==(const FullQName & other) const
```

Comment:
Overall, there are two problems here:

1. comparison must be done case-insensitively on A-Z/a-z for low 7-bit of UTF-8 strings  (see https://tools.ietf.org/html/rfc6762#section-16, which I quoted below)
2. There is no validation of UTF-8 stream validity (just basic sanity checking of "no unterminated sequences" should be done). The case insensitive matching mentioned above doesn't care about UTF-8 iteration, but there should be a pre-check at Qname extraction from query that the data is well-formed UTF-8.

>   The simple rules for case-insensitivity in Unicast DNS [RFC1034]
>    [RFC1035] also apply in Multicast DNS; that is to say, in name
>    comparisons, the lowercase letters "a" to "z" (0x61 to 0x7A) match
>    their uppercase equivalents "A" to "Z" (0x41 to 0x5A).  Hence, if a
>    querier issues a query for an address record with the name
>    "myprinter.local.", then a responder having an address record with
>    the name "MyPrinter.local." should issue a response.  No other
>    automatic equivalences should be assumed.  In particular, all UTF-8
>    multibyte characters (codes 0x80 and higher) are compared by simple
>    binary comparison of the raw byte values.  Accented characters are
>    *not* defined to be automatically equivalent to their unaccented
>    counterparts.  Where automatic equivalences are desired, this may be
>    achieved through the use of programmatically generated CNAME records.
>    For example, if a responder has an address record for an accented
>    name Y, and a querier issues a query for a name X, where X is the
>    same as Y with all the accents removed, then the responder may issue
>    a response containing two resource records: a CNAME record "X CNAME
>    Y", asserting that the requested name X (unaccented) is an alias for
>    the true (accented) name Y, followed by the address record for Y.

Reply:
Well, this is annoying. I will update for case insensitive stuff as well then.